src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,344 @@
+        }
+
+        /// <summary>
+        /// Returns all specializations that are used directly or indirectly within the given caller,
+        /// whether they are called, partially applied, or assigned. Each key in the returned dictionary
+        /// represents a specialization that is used by the caller. Each value in the dictionary is an
+        /// array of edges representing all the different ways the given caller specialization took a
+        /// dependency on the specialization represented by the associated key.
+        /// </summary>
+        public Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>> GetAllDependencies(QsSpecialization callerSpec) =>
+            GetAllDependencies(new CallGraphNode { CallableName = callerSpec.Parent, Kind = callerSpec.Kind, TypeArgs = RemovePositionFromTypeArgs(callerSpec.TypeArguments) });
+
+        /// <summary>
+        /// Finds and returns a list of all cycles in the call graph, each one being represented by an array of nodes.
+        /// To get the edges between the nodes of a given cycle, use the GetDirectDependencies method.
+        /// </summary>
+        public List<ImmutableArray<CallGraphNode>> GetCallCycles()
+        {
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
+                {
+                    foreach (var (curr, _) in dependencies)
+                    {
+                        if (!finished.Contains(curr))
+                        {
+                            if (callStack.TryGetValue(curr, out var next))
+                            {
+                                // Cycle detected
+                                var cycle = new List<CallGraphNode>() { curr };
+                                if (!curr.Equals(next)) // If the cycle is a direct recursion, we only want the one node
+                                {
+                                    do
+                                    {
+                                        cycle.Add(next);
+                                    } while (callStack.TryGetValue(next, out next));
+                                }
+
+                                cycles.Add(cycle.ToImmutableArray());
+                            }
+                            else
+                            {
+                                callStack[node] = curr;
+                                processDependencies(curr);
+                                callStack.Remove(node);
+                            }
+                        }
+                    }
+                }
+
+                finished.Add(node);
+            }
+
+            foreach (var key in _Dependencies.Keys)
```

Comment:
>var key [](start = 21, length = 7)

A comment here explaining that we are just iterating over all nodes in the graph would be useful for understanding. An even better comment would be to note that in reality we just need to iterate over entry points. Also if "var key" was "CallGraphNode node" I would read it much smoother. Why is it called a key? What is important about it, that it is a key in some dictionary or that it is a node in the graph?